metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1632,21 +1632,58 @@ boolean arePartitionLeadersImbalanced() {
     }
 
     /**
-     * Attempt to elect a preferred leader for all topic partitions which have a leader that is not the preferred replica.
+     * Check if we can do an election for partitions with no leader or a leader other than the preferred one.
      *
      * The response() method in the return object is true if this method returned without electing all possible preferred replicas.
      * The quorum controller should reschedule this operation immediately if it is true.
      *
      * @return All of the election records and if there may be more available preferred replicas to elect as leader
      */
-    ControllerResult<Boolean> maybeBalancePartitionLeaders() {
+    ControllerResult<Boolean> maybeAdjustPartitionLeaders() {
         List<ApiMessageAndVersion> records = new ArrayList<>();
+        maybeTriggerUncleanLeaderElectionForLeaderlessPartitions(records, maxElectionsPerImbalance);
+        maybeTriggerLeaderChangeForPartitionsWithoutPreferredLeader(records, maxElectionsPerImbalance);
+        boolean rescheduleImmediately = records.size() >= maxElectionsPerImbalance;
+        return ControllerResult.of(records, rescheduleImmediately);
+    }
 
-        boolean rescheduleImmediately = false;
+    /**
+     * Trigger unclean leader election for partitions without leader (visiable for testing)
+     *
+     * @param records  The record list to append to.
+     */
+    void maybeTriggerUncleanLeaderElectionForLeaderlessPartitions(
+        List<ApiMessageAndVersion> records,
+        int maxElections
+    ) {
+        Iterator<TopicIdPartition> iterator = brokersToIsrs.partitionsWithNoLeader();
+        for (TopicIdPartition topicIdPartition = iterator.next();
+             iterator.hasNext() && records.size() < maxElections; ) {
+            TopicControlInfo topic = topics.get(topicIdPartition.topicId());
+            if (configurationControl.uncleanLeaderElectionEnabledForTopic(topic.name)) {
+                ApiError result = electLeader(topic.name, topicIdPartition.partitionId(),
+                        ElectionType.UNCLEAN, records);
+                if (result.error().equals(Errors.NONE)) {
+                    log.error("Triggered unclean leader election for offline partition {}-{}.",
```

Comment:
Strictly speaking, the partition change records haven't yet been committed, so in rare circumstances, the wording in this log line will be misleading.

```suggestion
                    log.error("Triggering unclean leader election for offline partition {}-{}.",
```